maintainers: add script for checking unstable version strings

rhendric opened the pull request:
This script is mainly to address complaints that the new unstable version format policy makes it difficult for PR reviewers to verify that an unstable version is correct.

Try it out on some recent PRs:
```
maintainers/scripts/check-unstable-version.sh 240960
maintainers/scripts/check-unstable-version.sh 241201
maintainers/scripts/check-unstable-version.sh 241393
```

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 commented on maintainers/scripts/check-unstable-version.sh:
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p git nix
+
+set -o errexit -o nounset -o pipefail
```
```suggestion
set -eou pipefail
```

SuperSandro2000 commented on maintainers/scripts/check-unstable-version.sh:
```diff
@@ -0,0 +1,107 @@
+if [[ "$pr" ]]; then
+    mkWorktree
+    git fetch --quiet "$nixpkgsRemote" "pull/$pr/head"
+    mkGitWorktree FETCH_HEAD
+    cd "$worktree"
+    mkdir .src
```
Creating a new worktree takes quite a bit of time on my machines. It would probably be a good idea to have an option to just use the current worktree.

rhendric commented:
@SuperSandro2000 Feedback addressed.

fricklerhandwerk commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -224,6 +224,8 @@ There are a few naming guidelines:
 
 Example: Given a project had its latest releases `2.2` in November 2021, and `3.0` in January 2022, a commit authored on March 15, 2022 for an upcoming bugfix release `2.2.1` would have `version = "2.2-unstable-2022-03-15"`.
 
+(Reviewers: you can often check that a package's unstable version is correct, or get a suggestion for what to use instead, by running `maintainers/scripts/check-unstable-version.sh $PR_NUMBER`. This script isn't foolproof; common sense and this manual take precedence.)
```
Please link to that file on GitHub, otherwise one has to do manual steps to find out details.

rhendric committed:
maintainers: add script for checking unstable version strings

This script is mainly to address complaints that the new unstable
version format policy makes it difficult for PR reviewers to verify that
an unstable version is correct.
